### ISSUE TITLE
Add adjustable utilization threshold for re-writing packs during prune

### ISSUE THREAD

**Issue by teknynja** [restic/restic#1985]
Output of `restic version`
--------------------------

restic 0.9.2 compiled with go1.10.3 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

Currently during a "prune" operation, restic will re-write any pack that contains any unused data. This can trigger a lot of read/write operations as it has to read the data from every affected pack and then write that data back to the backend. This can become expensive on metered backends.

I would like to see an option to the prune command that allows the user to specify the amount of unused space allowed in a pack before it is re-written. Setting this option to 0% would cause packs with any amount of unused data to be re-written (the current behavior), whereas setting this option to 100% would only re-write packs that are completely unused. 

Note this feature was originally suggested as part of a broader discussion in issue [#1470 Parallelize prune pack rewrites](1470).

What are you trying to do?
--------------------------

Adding this feature would give the user some degree of control over how much read/write traffic is generated by a prune operation, with the trade-off being how much unused data is left behind in the repository.

Did restic help you or made you happy in any way?
-------------------------------------------------

I discovered restic like many others after CrashPlan discontinued their "Home" plan and I needed to find a suitable alternative. restic was the perfect fit, allowing me to choose different backend services as well as being a command line based utility, which allows me to run it easily on my headless server. I have been recommending it since to friends and co-workers. Thanks! :+1:

**Comment by teknynja** [comment]
@fd0 - Hi again! I've created a fork and I'm ready to get started on adding this feature.

First off, I'd like to ask your opinion about how to add the command flag for this. In my "spike", I added a `--repack-threshold` flag to both the `prune` _and_ the `forget` commands. This seemed a little awkward as I had some duplicate code in both modules, as well as having to create a PruneOptions variable in `cmd_forget.go`. The other option seems a bit problematic as well, which would be to add the `--repack-threshold` flag to the global options, as it really isn't global since it only affects two commands.

I wasn't able to find a mechanism in `spf13/cobra` that allowed parsing of a different command's flags withing another command, which could perhaps be an option here.

What are your thoughts on this? (Maybe there's another option I haven't even considered yet! :smirk:)

**Comment by fd0** [comment]
Hm, let me get back to you on that, I need to have a look. Either we need to build a function which adds the options (so that we can add them to `prune` and `forget`), or we define that such advanced options can only be passed when `prune` is run directly. That's a valid option in my opinion because it does not clutter the (already very complex) list of options for `forget`.

Let's leave `forget` aside and build the actual implementation, that's much more interesting :)

**Comment by teknynja** [comment]
Even if I don't add the flag to the `forget` command, I'll still need to create a `PruneOptions` instance in `cmd_forget.go` since I'll need to pass that object to the `pruneRepository` call which will now expect that parameter. In other words, since I need to pass in a `PruneOptions` object to `pruneRepository` when running the `prune` command, I'll also need to pass it in when invoking it via `forget`.

I can easily create an instance of `PruneOptions` configured with the old, default behavior in `cmd_forget.go`. At that point it's not a lot of effort to add the flag to `forget` as well, but like you said that command already has a lot of flags.

I guess for now I'll leave the flag off, but if you think having it there could be beneficial let me know and I'll toss it in while I'm in there.

**Comment by teknynja** [comment]
On the other hand, if in the future the `prune` command acquires additional flags, remembering to port them over to `forget` will become a maintenance issue - I think it makes sense to have `forget` invoke `prune` with just the defaults, and if a user needs to control the behavior of `prune` they'll just have to invoke it separately.

**Comment by teknynja** [comment]
Here is the proposed output from `restic help prune` (with the global flags removed):

```
The "prune" command checks the repository and removes data that is not
referenced and therefore not needed any more.

Usage:
  restic prune [flags]

Flags:
      --repack-threshold n   only rebuild packs with at least n% unused space (default: 0)
  -h, --help                 help for prune
```

Let me know if the flag name is ok and if the description verbiage needs adjustment.

**Comment by fd0** [comment]
Looks fine! I have an idea on how to generalize adding the flags for `prune` so that it can be reused in `forget`, pass a `PruneOptions` struct with the default values for now.

Also, I like to set the default value for the threshold to 20%. I wonder if we can come up with a better name...

**Comment by teknynja** [comment]
I worry about setting the default to something that changes the current behavior - users might wonder why prune is now leaving unused data behind when it didn't before. 0% gives them the current behavior (all unused data is removed). Judging from the discussion in #1470 it sounds like people that will use this flag will likely set it to 100% (only delete packs when they are completely unused). The cases for values in-between are a bit trickier because it depends on the type of data being stored (lots of large files, lots of small files, a mix of sizes, etc).

**Comment by fd0** [comment]
I think it's okay for this feature to be introduced with a slightly different behavior than before, as long as we have an entry in the CHANGELOG file :)

**Comment by teknynja** [comment]
Alrighty, 20% it is!

**Comment by teknynja** [comment]
So I've got the feature implemented in `cmd_prune.go`, along with some additional refactoring to remove some redundant loops and to move more data into the `stats` structure. I've added an entry in `changelog/unreleased`, and I've added the new prune flag to `doc/bash-completion.sh` (although I'm not sure how to test that yet).

Speaking of tests, they currently pass, but I notice that there's only integration tests to exercise the prune command, and I have no idea how to go about creating tests for the specific behavior of prune, let alone the behavior I've just added.

I also noticed that the documentation only touches lightly on the prune command, not going into specifics and really only mentioning that it should be run after `forget` or as the `--prune` option. I'm not sure if there's a place at that level of discussion to talk about the new prune flag.

Finally, I see that there are some kind of "man page" source files in the repository at `doc/man` but it looks like they may be generated somehow. I didn't look too hard into how that happens or why the results are in the repository, but I did not modify any of those files.

Basically, I'm working towards getting a PR together and I want to make sure I have any loose ends cleaned up. Let me know your thoughts on the above and if there's anything else I need to update for this feature.

Thanks!

**Comment by fd0** [comment]
Let's look at the code, can you create a PR so we can discuss it directly? Don't worry if it is not finished, we'll get there. :)

> and I've added the new prune flag to doc/bash-completion.sh (although I'm not sure how to test that yet).

Ah, hm. That's automatically generated from the CLI framework we're using (cobra), the same goes for the man pages. The `restic generate` command does that.

The main function is `repository.Repack` here: https://github.com/restic/restic/blob/master/internal/repository/repack.go#L20:6

Here are the tests for the function:
https://github.com/restic/restic/blob/master/internal/repository/repack_test.go

**Comment by teknynja** [comment]
Ooops! - Ok, I've reverted `bash-completion.sh` for now.

Created pull request #1994. See you there!

**Comment by fd0** [comment]
This issue is actually a duplicate of #348, closing the one over there.
